Visualize: Upgrading vis with pipeline agg from 6.8 doesn't render

VladLasitsa opened the pull request:
Closes: #93026

## Summary

As I understand start from 7.12 version we don't support 'schema' as object and should use the name of schema instead. Created migartion function find all schemes object in aggregation and replace it on name of this schema.

Schema before migration:
![image](https://user-images.githubusercontent.com/16915480/109939966-a2294980-7ce2-11eb-809b-c36631e37ecb.png)


After migration:
![image](https://user-images.githubusercontent.com/16915480/109940166-d6046f00-7ce2-11eb-8b0f-490f3ccc0c14.png)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/)
* Commit: ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-93427/ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/","number":"110729","commit":"ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989"}]}
PIPELINE-->

lukeelmers reviewed (commented):
The change here makes sense to me, but it might be good to also get @ppisljar to take a look since he did much of the schema refactoring.

alexwizp reviewed (approved):
Looks good to me! thank you

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

flash1293 commented:
Didn't do a deep dive but this export still doesn't render for me (with errors in the console): https://gist.github.com/flash1293/ba8a4cf238e16942d68161f1fee41541

flash1293 reviewed (changes requested):
Didn't do a deep dive but this export still doesn't render for me (with errors in the console): https://gist.github.com/flash1293/ba8a4cf238e16942d68161f1fee41541

VladLasitsa commented:
> Didn't do a deep dive but this export still doesn't render for me (with errors in the console): https://gist.github.com/flash1293/ba8a4cf238e16942d68161f1fee41541

I think it because this export already have version 7.12.

flash1293 reviewed (approved):
My bad @VladLasitsa , all my test cases work fine now, thanks!

stratoula commented:
@VladLasitsa this fails only when I upgrade from 6.8 to 7.12 right? It works for 7.11.x right? So this means that this bug doesn't exist yet (for the users). If this is the case, I would recommend replacing the `fix` label with the `skip` one. Unless this bug is live.

VladLasitsa commented:
@elasticmachine merge upstream

VladLasitsa commented:
> @VladLasitsa this fails only when I upgrade from 6.8 to 7.12? It works for 7.11.x right? So this means that this bug doesn't exist yet (for the users). If this is the case, I would recommend replacing the `fix` label with the `skip` one. Unless this bug is live.

Joe has written in issue that in 7.11 works fine. So in this case I will replace label.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111375/)
* Commit: c824535d775ebefe9a7db2fc3c2b98a2d121913d
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-93427/c824535d775ebefe9a7db2fc3c2b98a2d121913d)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #110729](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/) succeeded ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @VladLasitsa

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/","number":"110729","commit":"ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111375/","number":"111375","commit":"c824535d775ebefe9a7db2fc3c2b98a2d121913d"}]}
PIPELINE-->

ppisljar reviewed (approved):
code LGTM

stratoula commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111712/)
* Commit: 8a1e7268cd3105dbfae00249b8140ab9c2f87354
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-93427/8a1e7268cd3105dbfae00249b8140ab9c2f87354)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #111375](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111375/) succeeded c824535d775ebefe9a7db2fc3c2b98a2d121913d
* :green_heart: [Build #110729](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/) succeeded ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @VladLasitsa

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/110729/","number":"110729","commit":"ab12b0fd3f3e4c7aefce80871b6a01a9c1c03989"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111375/","number":"111375","commit":"c824535d775ebefe9a7db2fc3c2b98a2d121913d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/111712/","number":"111712","commit":"8a1e7268cd3105dbfae00249b8140ab9c2f87354"}]}
PIPELINE-->

stratoula commented:
@elasticmachine run elasticsearch-ci/docs

nodakai commented:
Hi, I have a strong suspicion that this matters to, at least, 7.11.1 as well. After we upgraded our Kibana installation to `kibana-7.11.1-1.x86_64`, some of our visualizations which we created long before broke and Firefox console showed something like
```
Uncaught (in promise) Error: Objects must have a type property
    getType expressions.plugin.js:4002
    getArgumentString expressions.plugin.js:5942
    getExpressionArgs expressions.plugin.js:5965
    getExpressionArgs expressions.plugin.js:5964
    getExpressionArgs expressions.plugin.js:5964
    getExpression expressions.plugin.js:5986
```

Saved objects -> Inspect shows `"schema": { ... }` under `customMetric` of the broken visualization

VladLasitsa commented:
> Hi, I have a strong suspicion that this matters to, at least, 7.11.1 as well. After we upgraded our Kibana installation to `kibana-7.11.1-1.x86_64`, some of our visualizations which we created long before broke and Firefox console showed something like
> 
> ```
> Uncaught (in promise) Error: Objects must have a type property
>     getType expressions.plugin.js:4002
>     getArgumentString expressions.plugin.js:5942
>     getExpressionArgs expressions.plugin.js:5965
>     getExpressionArgs expressions.plugin.js:5964
>     getExpressionArgs expressions.plugin.js:5964
>     getExpression expressions.plugin.js:5986
> ```
> 
> Saved objects -> Inspect shows `"schema": { ... }` under `customMetric` of the broken visualization

Hi @flash1293 , as you see @nodakai find out that in 7.11.1 old visualizations doesn't work due to problem which I fixed in this PR.  In the issue (https://github.com/elastic/kibana/issues/93026) you have written that in 7.11 old visualizations work fine. Because of this I have worked on this issue starting check from 7.12. As a result I added migration script for 7.12. Should we move migration script to 7.11 or should we leave as now and ask users modify visualizations manually?

nodakai commented:
I used the Inspect screen to search & destroy `schema` objects like the migration script would do, and that actually fixed our visualizations. Just FYI

flash1293 commented:
@VladLasitsa In the issue I wrote

> Upgrade from 7.11 works fine AFAICT

I tested upgrading a vis saved in 7.11 to 7.12 (I didn't test 6.8 to 7.11 - from 7.11, not to 7.11)

We can't change migration scripts for versions which are already released - the only option is to backport this migration to 7.11.2 . cc @stratoula not sure whether it's worth it, what do you think?

stratoula commented:
Yes, I don't think it worths it either and I think we can't backport it on 7.11.2. It is a bad bug, I will talk to @KOTungseth to see if we could at least include it in `the bugs solved` section. This is the reason I wanted to be sure here https://github.com/elastic/kibana/pull/93427#issuecomment-791395665
